### PR TITLE
Add `psnip_explicit_bzero` function for securely erasing a buffer

### LINKED ISSUE #39: Add a function for securely erasing a buffer
After using any sensitive data such as passwords, the buffers which store them must be cleared by the code to prevent it from being misused by a malicious actor by performing core dumps or using other means to access the memory of the application. However, most compilers optimize away the `memset` call as they interpret the call as dead code because the memory being written to is not being used afterwards.

I have already added a `psnip_memset_explicit` function in https://github.com/arnavyc/portable-snippets/tree/arnavyc/memset-explicit/memset-explicit that can erase a buffer without fear of 'dead-store removal'

### PR DESCRIPTION
**Pull request by ghost**
Closes #39.

Changes:
- Added version checks for Glibc, OpenBSD, NetBSD & FreeBSD.
- Added code to use `memset_s` C11 Annex K function, as it is guaranteed to not be optimized away.
- Added tests for testing `psnip_explicit_bzero` function.
- Use trick used by OpenSSL's [OPENSSL_cleanse](https://github.com/openssl/openssl/blob/f77208693ec3bda99618e6f76c0f8d279c0077bb/crypto/mem_clr.c) function to access `memset` using a `volatile` pointer, so that compiler does not optimize the function call.

### CODE REVIEW

File: explicit-bzero/explicit-bzero.c [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#endif /* _WIN32 */
+
+#ifdef __STDC_LIB_EXT1__
+/* memset_s() function in C11 Annex K (bound-checking interface) */
+#define HAVE_MEMSET_S 1
+#endif /* __STDC_LIB_EXT1__ */
+
+#if (defined(__GLIBC__) && __GLIBC_PREREQ(2, 25)) ||                           \
```
Comment (nemequ) [inline explicit-bzero/explicit-bzero.c]: The preprocessor doesn't short-circuit when evaluating conditions, so in this case the preprocessor will actually try to expand `__GLIBC_PREREQ(2, 25)` even if it doesn't exist, leading to an error on non-glibc platforms (I see this problem a lot with people trying to do stuff like `defined(__has_builtin) && __has_builtin(__builtin_memset)`).  You could do something like:

```c
#if defined(__GLIBC__)
  #if __GLIBC_PREREQ(2, 25)
    #define HAVE_EXPLICIT_BZERO
  #endif
#endif
#if !defined(HAVE_EXPLICIT_BZERO) && ( \
      (defined(__FreeBSD__) && __FreeBSD_version >= 1100037) || \
      (defined(__OpenBSD__) && OpenBSD >= 201405) \
    )
  #define HAVE_EXPLICIT_MEMSET
#endif
```

But that's pretty ugly.  I tend to prefer avoiding the macro altogether and just checking `__GLIBC__` and `__GLIBC_MINOR__`… the standard specifies that undefined object-like macros are replaced with 0, so it should be safe:

```c
#if \
    (defined(__GLIBC__) && (__GLIBC__ > 2 || (__GLIBC__ > 2 && __GLIBC_MINOR__ >= 25))) || \
    (defined(__FreeBSD__) && __FreeBSD_version >= 1100037) || \
    (defined(__OpenBSD__) && OpenBSD >= 201405) \
```

File: explicit-bzero/explicit-bzero.c [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+#endif
+
+#ifdef _WIN32
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#endif /* _WIN32 */
+
+#ifdef __STDC_LIB_EXT1__
+/* memset_s() function in C11 Annex K (bound-checking interface) */
+#define HAVE_MEMSET_S 1
```
Comment (nemequ) [inline explicit-bzero/explicit-bzero.c]: I'd rather move these checks into the function.  We only need to check `HAVE_MEMSET_S` once, so IMHO it's better to keep the logic near the code.

File: explicit-bzero/explicit-bzero.c [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  (void)explicit_memset(str, 0, n);
+#elif defined(HAVE_MEMSET_S)
+  (void)memset_s(str, (rsize_t)n, 0, (rsize_t)n);
+#elif defined(_WIN32)
+  (void)SecureZeroMemory(str, n);
+#elif defined(HAVE_EXPLICIT_BZERO)
+  explicit_bzero(str, n);
+#else
+
+#if defined(HAVE_WEAK_LINKING_SUPPORT)
```
Comment (nemequ) [inline explicit-bzero/explicit-bzero.c]: I don't see the benefit to this method for this code; the volatile pointer version does basically the same thing, but doesn't rely on weak linking being supported.

File: explicit-bzero/explicit-bzero.c [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+#ifndef __STDC_WANT_LIB_EXT1__
+/* memset_s() function in C11 Annex K (bound-checking interface) */
+#define __STDC_WANT_LIB_EXT1__ 1
+#endif
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "explicit-bzero.h"
```
Comment (nemequ) [inline explicit-bzero/explicit-bzero.c]: I'd prefer to not have a separate C file, and just move the entire implementation into the header.  If you move the implementation into the header than all you have to do to integrate this into your code is #include the header, whereas right now you also need to add the C file into your build system.

A separate C file is great when you have lots of code which would bloat the executable if you included a copy everywhere, or if you need to maintain global state, but in this case the implementation is so simple that I think a header-only implementation would be a much better way to go.

**Review by nemequ — CHANGES_REQUESTED** [review]
Thanks for working on this, and sorry it took me so long to get it reviewed ☹.  Overall, this looks excellent.  The checks seem solid, it's well documented and tested, and I like the interface.

My biggest concern is having it in a separate C file instead of just putting the implementation in a header.  Other than that, I've just added a few fairly minor suggestions.

If you're busy or don't particularly feel like changing the issues I mentioned, just let me know and I can make the changes when I merge.  You've clearly done the hard part, I don't mind taking care of the few little clean-up tasks remaining ;)

File: explicit-bzero/explicit-bzero.c (hunk 4 above)
Comment (ghost) [reply explicit-bzero/explicit-bzero.c]: I don't have any problem in doing so, however I cannot figure out a way to call `SecureZeroMemory` without including `windows.h`. Moreover, I am not using Windows so I can't do much about it.

Comment (nemequ) [reply explicit-bzero/explicit-bzero.c]: `SecureZeroMemory` is just a macro which is defined to `RtlSecureZeroMemory`.  `RtlSecureZeroMemory` doesn't need `<Windows.h>`, `<Wdm.h>` should include the necessary headers.  So, something like:

```c
#if defined(_WIN32) && !defined(SecureZeroMemory)
  #include <Wdm.h>
#endif

...

#elif defined(_WIN32)
  #if defined(SecureZeroMemory)
    SecureZeroMemory(...);
  #else
    RtlSecureZeroMemory(...);
  #endif
#endif
```

Comment (ghost) [reply explicit-bzero/explicit-bzero.c]: Done (see commit 81fc55f)

**Comment by pitrou** [comment]
I'm not sure why this PR was closed, but this function would still be useful.

**Pull request closed without merging**